Map annotation web

will-moore opened the pull request:
This PR adds Map Annotations to web.
Testing includes these areas:

 - Check that edits on the client are sync'd to the server with no data loss or overwriting (during copy, paste, edit, delete etc).
 - Check in DB that only a single map annotation per user per PDI is created.
 - Permissions testing - canAnnotate() allows adding of Map annotations, canEdit() allows editing of someone else's map annotation. TODO: removal of whole annotation.
 - Also good to compare clients and suggestions to improve UI & usability.

--no-rebase

![screen shot 2015-01-16 at 10 00 58](https://cloud.githubusercontent.com/assets/900055/5774516/aca1743a-9d66-11e4-840d-4ed1235faea8.png)

will-moore commented:
NB: To test the appearance of MapAnnotations added by the API, you can do...
```
$ bin/omero shell --login
from omero.gateway import BlitzGateway
conn = BlitzGateway(client_obj=client)
p = conn.getObject("Project", 1361)
ann = omero.gateway.MapAnnotationWrapper()
ann.setValue([["this is a","map Ann"],["added","via"],["the api", "with a"],["custom","namespace"]])
ann.setNs("some.value")
p.linkAnnotation(ann)
```

aleksandra-tarkowska commented:
@will-moore, it works really nice. Few comments:
 - toolbar, add ```+``` add button to add next row, it is good idea to add next row when click enter, but you may end up with a state where you do not edit anything and then what to click
 - toolbar, change order to: add, copy, past, delete
 - when click new row [*Add key*, *Add value*] default value should disappear from then input field then you will not end up with default value saved to DB
 - css style adjustment, see picture:
![google chromescreensnapz001](https://cloud.githubusercontent.com/assets/1065155/5775707/8a87d4fc-9d75-11e4-89c7-9b75af234167.png)
  - existing row font color ```#333```, new row font color ```#aaa``` and italic

aleksandra-tarkowska commented:
@dominikl would you like to apply similar changes to Insight?

will-moore commented:
Alternative styles for the 'placeholder' row. Any preferences? I think I prefer the white background (first image).

![screen shot 2015-01-16 at 12 00 21](https://cloud.githubusercontent.com/assets/900055/5775828/608ab64a-9d77-11e4-8c99-3e11771429e1.png)
![screen shot 2015-01-16 at 12 00 37](https://cloud.githubusercontent.com/assets/900055/5775829/608fa718-9d77-11e4-9e51-ca4c4d191222.png)

dominikl commented:
@will-moore  I'd also prefer the usual table row background color, the grey example looks like the row would be disabled.
@aleksandra-tarkowska I don't quite understand the purpose of the 'add row' button. What should happen if you click on it? There is already the 'Add key | Add value' row for adding new entries.

aleksandra-tarkowska commented:
@dominikl by default web do not add automatically new row when previous is saved and you can end up with no new entry (as I explained to you. Either Add button on automatically create new row when last one is saved. What ever you and @will-moore prefer.

manics commented:
If you enter a single `\` as a key or value it gets doubled: `\\`:

manics commented:
Also unicode characters aren't displayed after a refresh (e.g. £, §, ±)

mtbc commented:
Travis failure.

jburel commented:
restarted Travis

pwalczysko commented:
Bug: clicked on ``Add rows``

```
raceback (most recent call last):

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/django/core/handlers/base.py", line 114, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/decorators.py", line 452, in wrapped
    retval = f(request, *args, **kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/decorators.py", line 502, in wrapper
    context = f(request, *args, **kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webclient/views.py", line 1458, in annotate_map
    ann = conn.getObject("MapAnnotation", annId)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 2916, in getObject
    query, params, self.SERVICE_OPTS)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 4097, in __call__
    return self.handle_exception(e, *args, **kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omeroweb/webclient/webclient_gateway.py", line 1947, in handle_exception
    e, *args, **kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omero/gateway/__init__.py", line 4094, in __call__
    return self.f(*args, **kwargs)

  File "/opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/lib/python/omero_api_IQuery_ice.py", line 201, in findByQuery
    return _M_omero.api.IQuery._op_findByQuery.invoke(self, ((query, params), _ctx))

ApiUsageException: exception ::omero::ApiUsageException
{
    serverStackTrace = ome.conditions.ApiUsageException: Query named:

	select obj from Annotation obj join fetch obj.details.owner as owner join fetch obj.details.group join fetch obj.details.creationEvent

has returned more than one Object
findBy methods must return a single value.
Please try findAllBy methods for queries which return Lists.
	at ome.logic.QueryImpl.throwNonUnique(QueryImpl.java:376)
	at ome.logic.QueryImpl.findByQuery(QueryImpl.java:368)
	at sun.reflect.GeneratedMethodAccessor707.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy80.findByQuery(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor707.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy80.findByQuery(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor718.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:179)
	at ome.services.throttling.Callback.run(Callback.java:56)
	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:149)
	at ome.services.blitz.impl.QueryI.findByQuery_async(QueryI.java:92)
	at sun.reflect.GeneratedMethodAccessor717.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at omero.cmd.CallContext.invoke(CallContext.java:78)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy81.findByQuery_async(Unknown Source)
	at omero.api._IQueryTie.findByQuery_async(_IQueryTie.java:122)
	at omero.api._IQueryDisp.___findByQuery(_IQueryDisp.java:354)
	at omero.api._IQueryDisp.__dispatch(_IQueryDisp.java:520)
	at IceInternal.Incoming.invoke(Incoming.java:159)
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2357)
	at Ice.ConnectionI.dispatch(ConnectionI.java:1208)
	at Ice.ConnectionI.message(ConnectionI.java:1163)
	at IceInternal.ThreadPool.run(ThreadPool.java:302)
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:643)
	at java.lang.Thread.run(Thread.java:744)

    serverExceptionClass = ome.conditions.ApiUsageException
    message = Query named:

	select obj from Annotation obj join fetch obj.details.owner as owner join fetch obj.details.group join fetch obj.details.creationEvent

has returned more than one Object
findBy methods must return a single value.
Please try findAllBy methods for queries which return Lists.
}


<WSGIRequest
path:/merge/webclient/annotate_map/,
GET:<QueryDict: {}>,
POST:<QueryDict: {u'project': [u'1015'], u'mapAnnotation': [u'[]']}>,
COOKIES:{'_ga': 'GA1.2.483423307.1416590819',
 'csrftoken': 'CFDjgGaVwzFkmVwBxRDBt0OMRe1ClEx4',
 'phpbb3_a9mfv_k': '5adbdde5922531ec',
 'phpbb3_a9mfv_sid': 'f4a424618a3afc083ca8a2db9b29331b',
 'phpbb3_a9mfv_u': '2759',
 'sessionid': 'fsmo4b3w1a0w053wi7f2hh8ge6v7d5jq',
 'style_cookie': 'null'},
META:{'CONTENT_LENGTH': '33',
 'CONTENT_TYPE': 'application/x-www-form-urlencoded; charset=UTF-8',
 u'CSRF_COOKIE': u'CFDjgGaVwzFkmVwBxRDBt0OMRe1ClEx4',
 'HTTPS': '',
 'HTTP_ACCEPT': '*/*',
 'HTTP_ACCEPT_ENCODING': 'gzip, deflate',
 'HTTP_ACCEPT_LANGUAGE': 'en-GB,en;q=0.5',
 'HTTP_CACHE_CONTROL': 'no-cache',
 'HTTP_CONNECTION': 'keep-alive',
 'HTTP_CONTENT_LENGTH': '33',
 'HTTP_CONTENT_TYPE': 'application/x-www-form-urlencoded; charset=UTF-8',
 'HTTP_COOKIE': 'phpbb3_a9mfv_u=2759; phpbb3_a9mfv_k=5adbdde5922531ec; phpbb3_a9mfv_sid=f4a424618a3afc083ca8a2db9b29331b; style_cookie=null; csrftoken=CFDjgGaVwzFkmVwBxRDBt0OMRe1ClEx4; _ga=GA1.2.483423307.1416590819; sessionid=fsmo4b3w1a0w053wi7f2hh8ge6v7d5jq',
 'HTTP_HOST': 'trout.openmicroscopy.org',
 'HTTP_PRAGMA': 'no-cache',
 'HTTP_REFERER': 'http://trout.openmicroscopy.org/merge/webclient/',
 'HTTP_USER_AGENT': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:33.0) Gecko/20100101 Firefox/33.0',
 'HTTP_X_CSRFTOKEN': 'CFDjgGaVwzFkmVwBxRDBt0OMRe1ClEx4',
 'HTTP_X_REQUESTED_WITH': 'XMLHttpRequest',
 'PATH_INFO': u'/webclient/annotate_map/',
 'QUERY_STRING': '',
 'REQUEST_METHOD': 'POST',
 'SCRIPT_INFO': '/merge',
 'SCRIPT_NAME': u'/merge',
 'SERVER_NAME': 'ome-c6100-3',
 'SERVER_PORT': '80',
 'SERVER_PROTOCOL': 'HTTP/1.1',
 'wsgi.errors': <flup.server.fcgi_base.TeeOutputStream object at 0x5a8f210>,
 'wsgi.input': <flup.server.fcgi_base.InputStream object at 0x5ba1490>,
 'wsgi.multiprocess': True,
 'wsgi.multithread': False,
 'wsgi.run_once': False,
 'wsgi.url_scheme': 'http',
 'wsgi.version': (1, 0)}>`
```

will-moore commented:
Petr also got the same exception as above when clicking on "Remove rows"
```
POST:<QueryDict: {u'mapAnnotation': [u'[]'], u'image': [u'7506']}>,
...
findBy methods must return a single value.
```

joshmoore commented:
@will-moore : is this because the assumption is that there will only be one "/client" map annotation?

will-moore commented:
@joshmoore Sorry - what do you mean by ```this```?
There should only be one "/client" map annotation per user per P/D/I etc.

joshmoore commented:
Sure, but nothing guarantees that.

dominikl commented:
Works like expected. One very minor flaw: The top of the scrollbar is hidden behind the toolbar, maybe it's possible to move the toolbar a few pixels to the left?
![screen shot 2015-01-21 at 09 54 30](https://cloud.githubusercontent.com/assets/6575139/5834301/f2c691d4-a153-11e4-90a9-ee49a8de82b3.png)

will-moore commented:
Now showing namespaces for non-client map annotations if set:

![screen shot 2015-01-21 at 11 02 50](https://cloud.githubusercontent.com/assets/900055/5835166/504da5f0-a15d-11e4-9772-3992e73ab08b.png)

manics commented:
Comments from https://github.com/openmicroscopy/openmicroscopy/pull/3333#issuecomment-70254652 and https://github.com/openmicroscopy/openmicroscopy/pull/3333#issuecomment-70260733 still apply

mtbc commented:
Not in merge. Don't know if Travis needs some kind of kick.

sbesson commented:
" We can’t automatically merge this pull request. " needs origin/develop merged in first.

will-moore commented:
@manics Thanks - hadn't forgotten. Should be fixed now.

dominikl commented:
Works like expected

jburel commented:
@will-moore: I assume you have more commits to come following last week discussion
e.g. If I click on + when the table is empty, the rows are added above the empty "key-value" row

will-moore commented on components/tools/OmeroWeb/omeroweb/webgateway/static/webgateway/js/ome.popup.js:
```diff
@@ -166,7 +166,7 @@ String.prototype.escapeHTML = function(){
         switch(s) {
             case "&": return "&amp;";
-            case "\\": return "\\\\";
+            // case "\\": return "\\\\";
```
@aleksandra-tarkowska Since you added this escapeHTML() function, you OK with removing this line from it?
I can't see that it's doing any good, but I might have missed something.

aleksandra-tarkowska replied:
just replace ```\\``` with ``` &#92; &#92;```

aleksandra-tarkowska replied:
This line doesn't escape single backslash, there must be something else in the code

manics commented:
Bit of a corner case, but....

1. Add a row with `<script>alert("hello")</script>`:`<script>alert("hello")</script>`.
2. Click on the copy icon
3. Paste... you'll get an alert dialog

manics commented:
Also clicking Add(+) multiple times without entering anything adds blank rows (which are discarded on refresh), it would be nicer to just not add additional rows.

manics commented:
Adding (`Add Key`, `Add Value`) (along with other keys) to the client annotation via the API works, but if you try to edit that annotaiton in the web client (`Add Key`, `Add Value`) is lost.

will-moore commented:
@manics I've fixed the paste escape HTML bug.
I think that adding multiple empty rows is actually preferable to disabling this behaviour. If the user wants to add multiple empty rows, then it is better to let them than to disable this and have to indicate why it doesn't work in a tool-tip or pop-up.
I've reverted back to saving 'Add Key', 'Add Value' rows instead of ignoring them. This works OK now as we remove that placeholder text more pro-actively than we did before.

Also updated escapeHTML() as suggested by @aleksandra-tarkowska.

will-moore commented:
@pwalczysko Do we need a full permissions test of Map Annotations before we can merge?

pwalczysko commented:
@will-moore : I was thinking about a test in size of the permissions testing pre-christmas. Will prepare something.

will-moore commented:
@pwalczysko @jburel On second thoughts, maybe we should merge this as is, then fix any outstanding permissions issues in another PR. That would help keep it a bit clearer since this is already a big PR. We have integration tests and robot tests that are passing.

pwalczysko commented:
Agree with @will-moore : lets keep the PRs small. Permissions testing will necessarily get us into a lot of organizing and confusion.

dominikl commented:
I'd suggest the same for the corresponding Insight PR #3338

jburel commented:
As discussed today, I will do a final check to make sure there is no blocker
We can then fix problems if any in follow-up PRs

joshmoore commented:
Merging then. Thanks all.

jburel commented:
Still few UI things to fix before the release, but it certainly usable and getting closer
I will add comment to Trello card.
Merging
